readme.txt
=== Single Category Permalink ===
Contributors: coffee2code
Donate link: https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522
Tags: permalink, structure, link, category, coffee2code
License: GPLv2 or later
License URI: https://www.gnu.org/licenses/gpl-2.0.html
Requires at least: 4.6
Tested up to: 5.4
Stable tag: 2.4.2

Reduce permalinks (category or post) that include entire hierarchy of categories to just having the lowest level category.


== Description ==

Reduce permalinks (category or post) that include entire hierarchy of categories to just having the lowest category in the hierarchy.

By default, WordPress replaces the %category% permalink tag in a custom permalink structure with the entire hierarchy of categories for the post's first matching category. For example, assuming your site has a hierarchical category structure like so:

`
Applications
  |_ Desktop
  |_ Web
    |_ WordPress
`

By default, if you have a permalink structure defined as `%category%/%year%/%monthnum%/%day%/%postname%`, your post titled "Best Plugins" assigned to the "WordPress" category would have a permalink of:

`https://www.example.com/applications/web/wordpress/2008/01/15/best-plugins`

If you activate the Single Category Permalink plugin, this would be the permalink generated for the post (and recognized by the blog):

`https://www.example.com/wordpress/2008/01/15/best-plugins`

In order for a category to be used as part of a post's permalink structure, `%category%` must be explicitly defined in the Settings -> Permalinks admin page as part of a custom structure, i.e. `/%category%/%postname%`.

For category links, `%category%` is implied to follow the value set as the "Category base" (or the default category base if none is specified). So if your category base is 'category', the above example would list posts in the 'WordPress' category on this category listing page:

`https://www.example.com/category/applications/web/wordpress/`

With this plugin activated, that link would become:

`https://www.example.com/category/wordpress/`

NOTE: The fully hierarchical category and post permalinks will continue to work. The plugin issues a 302 redirect to browsers and search engines pointing them to the shorter URL.

Links: [Plugin Homepage](https://coffee2code.com/wp-plugins/single-category-permalink/) | [Plugin Directory Page](https://wordpress.org/plugins/single-category-permalink/) | [GitHub](https://github.com/coffee2code/single-category-permalink/) | [Author Homepage](https://coffee2code.com)


== Installation ==

1. Install via the built-in WordPress plugin installer. Or download and unzip `single-category-permalinks.zip` inside the plugins directory for your site (typically `wp-content/plugins/`)
2. Activate the plugin through the 'Plugins' admin menu in WordPress
3. Use `%category%` as a permalink tag in the `Settings` -> `Permalinks` admin options page when defining a custom permalink structure


== Frequently Asked Questions ==

= Will existing links to my site that used the full category hierarchy still work? =

Yes, WordPress will still serve the category listings and posts regardless of whether it is of the full category hierarchy format or just the single category format. But do note that WordPress doesn't perform any sort of redirects; it responds directly to the category/post URL requested.

= Could this give the appearance that I have duplicate content on my site if pages are accessible via the full category hierarchy permalink format and the single category permalink format? =

Whether this plugin is active or not, WordPress treats both types of category links the same. This plugin will however issue redirects for all of the non-canonical category and post permalink pages to point to the single category link version.

= What can this plugin do for me if I don't use `%category%` in my custom permalink structure? =

In addition to handling custom permalink structures (used to generate permalinks for posts) that contain `%category%`, the plugin also shortens category archive links. WordPress by default generates those links in a fully hierarchical fashion which this plugin will reduce to a single category. See the Description section for an example.

= Does this plugin include unit tests? =

Yes.


== Hooks ==

The plugin exposes one filter for hooking. Code using this filter should ideally be put into a mu-plugin or site-specific plugin (which is beyond the scope of this readme to explain). Less ideally, you could put them in your active theme's functions.php file.

**c2c_single_category_redirect_status (filter)**

The 'c2c_single_category_redirect_status' hook allows you to specify an HTTP status code used for the redirect. By default this is 302.

Arguments:

* $status (integer) : The default HTTP status code

Example:

`
/**
 * Change the redirection HTTP status to a 302.
 *
 * @param  int $code The HTTP status code. By default 301.
 * @return int
 */
function scp_change_redirect_status( $code ) {
	return 302;
}
add_filter( 'c2c_single_category_redirect_status', 'scp_change_redirect_status' );
`


== Changelog ==

= 3.0.1 (2020-08-21) =
* Change: fixed the sitemap of Yoast

= 3.0 (2020-08-04) =
* Change: The primary category exists ONLY in Yoast, it's not a WP native property. Now it's handled the behaviour in case of multiple categories, primary category defined or not
* Change: added cache

= 2.6 (2020-08-01) =
* Change: Now the plugin works also with Yoast (changes the canonical url, the og:url, and the url inside the schema)

= 2.5 (2020-08-01) =
* Change: Fixed problems when a post is under 2 hierarchy of categories (one primary and one not)

= 2.4.2 (2020-06-01) =
* Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests (and remove commented-out code)
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* Change: Update URLs used in examples to be HTTPS
* Change: Unit tests: Remove unnecessary unregistering of hooks

= 2.4.1 (2020-01-18) =
* Fix: Prevent interfering with permalinks for non-category taxonomies. Props mynameisrabby.

= 2.4 (2020-01-17) =
* Change: Use `term_link` filter instead of the deprecated `category_link` filter. Props webgmclassics.
* New: Add TODO.md and move existing TODO list from top of main plugin file into it (and add more items to the list)
* Unit tests:
    * Change: Update unit test install script and bootstrap to use latest WP unit test repo
    * New: Add test for default hooks
    * Fix: Rename and fix miscoded test for plugin initialization
* Change: Note compatibility through WP 5.3+
* Change: Update copyright date (2020)
* Fix: Correct typo in GitHub URL

_Full changelog is available in [CHANGELOG.md](https://github.com/coffee2code/single-category-permalink/blob/master/CHANGELOG.md)._


== Upgrade Notice ==

= 2.4.2 =
Trivial update: Updated a few URLs to be HTTPS and noted compatibility through WP 5.4+

= 2.4.1 =
Recommended bugfix release: prevented interfering with permalinks for non-category taxonomies

= 2.4 =
Minor update: modernized unit tests, noted compatibility through WP 5.3+, added TODO.md, and updated copyright date (2020)

= 2.3 =
Minor update: tweaked plugin initialization process, created CHANGELOG.md to store historical changelog outside of readme.txt, noted compatibility through WP 5.1+, updated copyright date (2019)

= 2.2.1 =
Trivial update: minor documentation and code formatting tweaks, noted compatibility through WP 4.9+, and updated copyright date (2018)

= 2.2 =
Recommended update: changed default HTTP redirect status code to 301 (permanent), fixed PHP warning in WP 4.7 due to function deprecation, restructured code, compatibility is now WP 4.6-4.7+, added more unit tests, updated copyright date, more

= 2.1.2 =
Trivial update: verified compatibility through WP 4.5; updated copyright date (2016).

= 2.1.1 =
Trivial update: noted compatibility through WP 4.1+; updated copyright date (2015); added plugin icon

= 2.1 =
Minor update: added unit tests; noted compatibility through WP 3.8+

= 2.0.4 =
Trivial update: noted compatibility through WP 3.5+

= 2.0.3 =
Trivial update: noted compatibility through WP 3.4+; explicitly stated license

= 2.0.2 =
Trivial update: noted compatibility through WP 3.3+

= 2.0.1 =
Bugfix release: fixed bug triggered when creating new post (especially recommended if using %category% in custom permalink structure)

= 2.0 =
Recommended update. No functional changes, but many changes to formatting and documentation; noted compatibility through WP 3.2.


CHANGELOG.md
# Changelog

# 3.0.1 _(2020-08-21)_
* Change: fixed the sitemap of Yoast

# 3.0 _(2020-08-04)_
* Change: The primary category exists ONLY in Yoast, it's not a WP native property. Now it's handled the behaviour in case of multiple categories, primary category defined or not
* Change: added cache

## 2.6 _(2020-08-01)_
* Change: Now the plugin works also with Yoast (changes the canonical url, the og:url, and the url inside the schema)

## 2.5 _(2020-08-01)_
* Change: Fixed problems when a post is under 2 hierarchy of categories (one primary and one not)

## 2.4.2 _(2020-06-01)_
* Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests (and remove commented-out code)
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* Change: Update URLs used in examples to be HTTPS
* Change: Unit tests: Remove unnecessary unregistering of hooks

## 2.4.1 _(2020-01-18)_
* Fix: Prevent interfering with permalinks for non-category taxonomies. Props mynameisrabby.

## 2.4 _(2020-01-17)_
* Change: Use `term_link` filter instead of the deprecated `category_link` filter. Props webgmclassics.
* New: Add TODO.md and move existing TODO list from top of main plugin file into it (and add more items to the list)
* Unit tests:
    * Change: Update unit test install script and bootstrap to use latest WP unit test repo
    * New: Add test for default hooks
    * Fix: Rename and fix miscoded test for plugin initialization
* Change: Note compatibility through WP 5.3+
* Change: Update copyright date (2020)
* Fix: Correct typo in GitHub URL

## 2.3 _(2019-03-18)_
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* New: Add inline documentation for hook
* Change: Initialize plugin on 'plugins_loaded' action instead of on load
* Change: Merge `do_init()` into `init()`
* CHange: Cast return value of `c2c_single_category_redirect_status` filter as integer
* Change: Split paragraph in README.md's "Support" section into two
* Change: Note compatibility through WP 5.1+
* Change: Update copyright date (2019)
* Change: Update License URI to be HTTPS

## 2.2.1 _(2018-07-21)_
* Change: Switch away from using deprecated `c2c_single_category_catlink()` internally
* Change: Move `@uses` DockBlock entries to `get_http_redirect_status()`
* New: Unit tests: Add `create_hierarchical_categories()` for creating hierarchical categories so the behavior isn't duplicated in a number of tests
* New: Add README.md
* New: Add GitHub link to readme
* Change: Minor whitespace tweaks to unit test bootstrap
* Change: Note compatibility through WP 4.9+
* Change: Rename readme.txt section from 'Filters' to 'Hooks'
* Change: Modify formatting of hook name in readme to prevent being uppercased when shown in the Plugin Directory
* Change: Improve installation instructions
* Change: Minor readme.txt tweaks
* Change: Update copyright date (2018)

## 2.2 _(2017-02-10)_
* Fix: Replace use of deprecated (in WP 4.7) `_usort_terms_by_ID()` with `wp_list_sort()` for WP 4.7+.
* Change: Change default redirect HTTP status code from 302 (temporary redirect) to 301 (permanent redirect).
* Change: Wrap functionality in class.
    * Create class `c2c_SingleCategoryPermalink`
    * Deprecate existing functions: `c2c_single_category_catlink()`, `c2c_single_category_postlink()`, `c2c_single_category_redirect()`
    * Move deprecated function functionality to class methods: `category_link()`, `post_link()`, `template_redirect()`
    * Add method `version()` to return plugin version
* Add `get_http_redirect_status()` for getting the HTTP status code for redirects.
* Change: Default `WP_TESTS_DIR` to `/tmp/wordpress-tests-lib` rather than erroring out if not defined via environment variable.
* Change: Enable more error output for unit tests.
* Change: Note compatibility through WP 4.7+.
* Change: Remove support for WordPress older than 4.6 (should still work for earlier versions back to WP 1.5)
* Change: Update copyright date (2017).

## 2.1.2 _(2016-03-29)_
* New: Add LICENSE file.
* New: Add empty index.php to prevent files from being listed if web server has enabled directory listings.
* New: Add 'Text Domain' to plugin header.
* Change: Minor code reformatting.
* Change: Explicitly declare methods in unit tests as public; minor unit test doc reformatting.
* Change: Note compatibility through WP 4.5+.
* Change: Update copyright date (2016).

## 2.1.1 _(2015-02-17)_
* Reformat plugin header
* Note compatibility through WP 4.1+
* Change documentation links to wp.org to be https
* Update copyright date (2015)
* Add plugin icon

## 2.1 _(2014-01-24)_
* Add unit tests
* Minor documentation improvements
* Minor code reformatting (spacing, bracing)
* Note compatibility through WP 3.8+
* Update copyright date (2014)
* Change donate link
* Add banner

## 2.0.4
* Add check to prevent execution of code if file is directly accessed
* Note compatibility through WP 3.5+
* Update copyright date (2013)
* Minor code reformatting (spacing)

## 2.0.3
* Re-license as GPLv2 or later (from X11)
* Add 'License' and 'License URI' header tags to readme.txt and plugin file
* Remove ending PHP close tag
* Note compatibility through WP 3.4+

## 2.0.2
* Note compatibility through WP 3.3+
* Add link to plugin directory page to readme.txt
* Update copyright date (2012)

## 2.0.1
* Fix bug triggered when creating new post

## 2.0
* Fix compatibility bug relating to generation of category permalink
* Rename `single_category_postlink()` to `c2c_single_category_postlink()`
* Rename `single_category_catlink()` to `c2c_single_category_catlink()`
* Add `c2c_single_category_redirect()` to redirect hierarchical category links to the single category alternative
* Add filter `c2c_single_category_redirect_status` to allow override of default redirect status code
* Wrap all functions in `if (!function_exists())` check
* Remove docs from top of plugin file (all that and more are in readme.txt)

tests/test-single-category-permalink.php
<?php

defined( 'ABSPATH' ) or die();

class Single_Category_Permalink_Test extends WP_UnitTestCase {

	public function setUp() {
		parent::setUp();
		$this->set_permalink();
	}


	//
	//
	// HELPER FUNCTIONS
	//
	//


	private function set_permalink( $structure = '/%category%/%postname%/', $category_structure = '/category/%category%/' ) {
		update_option( 'permalink_structure', $structure );
		update_option( 'category_base',       $category_structure );
		$GLOBALS['wp_rewrite']->init();
		add_permastruct( 'category', ltrim( $category_structure, '/' ) );
		flush_rewrite_rules();
	}

	public function change_redirect_status( $status ) {
		return 302;
	}

	private function create_hierarchical_categories() {
		$cat1_id = $this->factory->category->create( array ( 'slug' => 'aaa', 'name' => 'AAA' ) );
		$cat2_id = $this->factory->category->create( array ( 'slug' => 'bbb', 'name' => 'BBB', 'parent' => $cat1_id ) );
		$cat3_id = $this->factory->category->create( array ( 'slug' => 'ccc', 'name' => 'CCC', 'parent' => $cat2_id ) );

		return array( $cat1_id, $cat2_id, $cat3_id );
	}

	//
	//
	// TESTS
	//
	//


	public function test_class_exists() {
		$this->assertTrue( class_exists( 'c2c_SingleCategoryPermalink' ) );
	}

	public function test_get_version() {
		$this->assertEquals( '2.5', c2c_SingleCategoryPermalink::version() );
	}

	public function test_hooks_plugins_loaded() {
		$this->assertEquals( 10, has_action( 'plugins_loaded', array( 'c2c_SingleCategoryPermalink', 'init' ) ) );
	}

	public function test_default_hooks() {
		$this->assertEquals( 10, has_filter( 'term_link',         array( 'c2c_SingleCategoryPermalink', 'category_link' ) ) );
		$this->assertEquals( 10, has_filter( 'post_link',         array( 'c2c_SingleCategoryPermalink', 'post_link' ) ) );
		$this->assertEquals( 10, has_filter( 'template_redirect', array( 'c2c_SingleCategoryPermalink', 'template_redirect' ) ) );
	}

	/* Test post permalink */

	public function test_non_hierarchical_category_not_affected() {
		$cat_id  = $this->factory->category->create( array ( 'slug' => 'aaa', 'name' => 'AAA' ) );
		$post_id = $this->factory->post->create( array( 'post_title' => 'Cat Post' ) );
		wp_set_post_categories( $post_id, $cat_id );

		$this->assertEquals( 'http://example.org/aaa/cat-post/', get_permalink( $post_id ) );
	}

	public function test_post_assigned_hierarchical_leaf_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();
		$post_id = $this->factory->post->create( array( 'post_title' => 'Cat Post' ) );
		wp_set_post_categories( $post_id, $cat3_id );

		$this->assertEquals( 'http://example.org/ccc/cat-post/', get_permalink( $post_id ) );
	}

	public function test_post_assigned_hierarchical_root_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();
		$post_id = $this->factory->post->create( array( 'post_title' => 'Cat Post' ) );
		wp_set_post_categories( $post_id, $cat1_id );

		$this->assertEquals( 'http://example.org/aaa/cat-post/', get_permalink( $post_id ) );
	}

	public function test_post_assigned_hierarchical_midlevel_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();
		$post_id = $this->factory->post->create( array( 'post_title' => 'Cat Post' ) );
		wp_set_post_categories( $post_id, $cat2_id );

		$this->assertEquals( 'http://example.org/bbb/cat-post/', get_permalink( $post_id ) );
	}

	public function test_post_assigned_multiple_hierarchical_leaf_categories() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();
		$cat4_id = $this->factory->category->create( array ( 'slug' => 'ddd', 'name' => 'DDD' ) );
		$post_id = $this->factory->post->create( array( 'post_title' => 'Cat Post' ) );
		wp_set_post_categories( $post_id, array( $cat3_id, $cat4_id ) );

		$this->assertEquals( 'http://example.org/ccc/cat-post/', get_permalink( $post_id ) );
	}

	/* Test category permalink */

	public function test_non_category_hierarchical_taxonomy_not_affected() {
		register_taxonomy( 'color', 'post', array(
			'hierarchical' => true,
			'has_archive'  => true,
			'rewrite'      => array( 'hierarchical' => true ),
		) );

		$term1_id = $this->factory->term->create( array ( 'taxonomy' => 'color', 'slug' => 'aaa', 'name' => 'AAA' ) );
		$term2_id = $this->factory->term->create( array ( 'taxonomy' => 'color', 'slug' => 'bbb', 'name' => 'BBB', 'parent' => $term1_id ) );
		$term3_id = $this->factory->term->create( array ( 'taxonomy' => 'color', 'slug' => 'ccc', 'name' => 'CCC', 'parent' => $term2_id ) );

		$this->assertEquals( 'http://example.org/color/aaa/bbb/ccc/', get_term_link( $term3_id ) );
	}

	public function test_category_permalink_for_leaf_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();

		$this->assertEquals( 'http://example.org/category/ccc/', get_category_link( $cat3_id ) );
	}

	public function test_category_permalink_for_root_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();

		$this->assertEquals( 'http://example.org/category/aaa/', get_category_link( $cat1_id ) );
	}

	public function test_category_permalink_for_midlevel_category() {
		list( $cat1_id, $cat2_id, $cat3_id ) = $this->create_hierarchical_categories();

		$this->assertEquals( 'http://example.org/category/bbb/', get_category_link( $cat2_id ) );
	}

	/*
	 * get_http_redirect_status()
	 */

	public function test_get_http_redirect_status() {
		$this->assertEquals( 301, c2c_SingleCategoryPermalink::get_http_redirect_status() );
	}

	/*
	 * Filter: c2c_single_category_redirect_status
	 */

	public function test_default_redirect_status() {
		add_filter( 'c2c_single_category_redirect_status', array( $this, 'change_redirect_status' ) );

		$this->assertEquals( 302, c2c_SingleCategoryPermalink::get_http_redirect_status() );
	}


	/* TODO: Test redirect of full hierarchical category permalink (post) to shorter version (e.g. /aaa/bbb/ccc/cat-post/ -> /ccc/cat-post/) */
	/* TODO: Test redirect of full hierarchical category permalink (category) to shorter version (e.g. /category/aaa/bbb/ccc/ -> /category/ccc/) */
	/* TODO: Test appropriate post(s) returned when querying shorter category permalink (e.g. /category/ccc/) */
	/* TODO: Test appropriate post(s) returned when querying leaf category off home (e.g. '/ccc/') */

}


single-category-permalink.php
<?php
/**
 * Plugin Name: Single Category Permalink
 * Version:     3.0.1
 * Plugin URI:  https://coffee2code.com/wp-plugins/single-category-permalink/
 * Author:      Scott Reilly, Gianluigi Salvi
 * Author URI:  https://coffee2code.com/
 * Text Domain: single-category-permalink
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Reduce permalinks (category or post) that include entire hierarchy of categories to just having the lowest level category.
 *
 * Compatible with WordPress 4.6 through 5.4+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/single-category-permalinks/
 *
 * @package Single_Category_Permalink
 * @author  Scott Reilly, Gianluigi Salvi
 * @version 3.0.1
 */

/*
	Copyright (c) 2007-2020 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined('ABSPATH') or exit();


class c2c_SingleCategoryPermalink
{

    /**
     * Returns version of the plugin.
     *
     * @since 2.2
     */
    public static function version()
    {
        return '2.6';
    }

    /**
     * Initialization.
     *
     * @since 2.2
     */
    public static function init()
    {
        // Load textdomain.
        load_plugin_textdomain('single-category-permalink');

        add_filter('term_link',             array(__CLASS__, 'category_link'),              1, 3);
        add_filter('post_link',             array(__CLASS__, 'post_link'),                  1, 2);
        add_filter('template_redirect',     array(__CLASS__, 'template_redirect'));
        add_filter('wpseo_canonical',       array(__CLASS__, 'yoast_change_seo_canonical'), 1, 1 );
        add_filter('wpseo_opengraph_url',   array(__CLASS__, 'yoast_change_seo_opengraph_url'), 10);
        add_filter('wpseo_schema_article',  array(__CLASS__, 'example_change_article') );
        add_filter('yoast_seo_development_mode', '__return_true' );
        //remove_action('wp_head',
        //add_action( 'wp_head', [ $this, 'call_wpseo_head' ], 1 );
    }

    /**
     * Changes
     * @param array $data Schema.org Article data array.
     * @return void Schema.org Article data array.
     */
    public static function example_change_article( $data )
    {
        //self::PrintStackTrace();

        //YoastSEO()->context_memoizer;

        //PC::debug(YoastSEO()->meta->for_current_page());
        //PC::debug($data);

        //Change wp-content\plugins\wordpress-seo-premium\src\presenters\canonical-presenter.php\
        //Add $this->presentation->canonical = get_permalink();

        YoastSEO()->meta->for_current_page()->canonical = get_permalink();
        //YoastSEO()->meta->context_memoizer->for_current_page()->canonical = get_permalink();
    }

    public static function yoast_change_seo_canonical( $canonical )
    {
        if ( is_single() )
        {
            return get_permalink();
        }
        else
            return $canonical;
    }

    public static function yoast_change_seo_opengraph_url( $canonical )
    {
        return get_permalink();
    }

    /**
     * Returns post URI for a given post.
     *
     * If the post permalink structure includes %category%, then this function
     * kicks into gear to reduce a hierarchical category structure to its lowest
     * category.
     *
     * @param string $permalink The default URI for the post
     * @param WP_Post $post The post
     * @return string             The post URI.
     */
    public static function post_link($permalink, $post)
    {
//        $redis = new Redis();
//        $redis->connect('127.0.0.1', 12345);
//        $redis->auth('MyPassword');

//        echo "Server is running: ".$redis->ping();

        $cachedRedirect = wp_cache_get("$permalink");
        if($cachedRedirect)
        {
//            PC::debug($cachedRedirect);
            return $cachedRedirect;
        }


        $permalink_structure = get_option('permalink_structure');

        // Only do anything if '%category%' is part of the post permalink
        if (strpos($permalink_structure, '%category%') !== false)
        {
            $categoryOtherMethod = self::GetLastCategory($post);
            $category = self::get_post_primary_category($post);

//            PC::debug($categoryOtherMethod);
//            PC::debug($category);

            // Find category hierachy for the category. By default, these would be part of the full category permalink.
            $category_hierarchy = $category->slug;
 //           PC::debug($category->parent);

            if ($parent = $category->parent /* the id of the parent category or 0*/)
            {
                $category_hierarchy = get_category_parents($parent, false, '/', true) . $category->slug;
                //medicina-e-salute/integratori
                //PC::debug("entrato");
            }

            // Now that the permalink component involving category hierarchy consists of is known,
            // replace it with the main category slug
            //PC::debug($category_hierarchy);
            //PC::debug($category->slug);

            $oldPermalink = $permalink;
            $permalink = str_replace($category_hierarchy, $category->slug, $permalink);
            wp_cache_set($oldPermalink, $permalink);
            //PC::debug($oldPermalink);
            //debug($permalink);
        }

        //PC::debug($permalink);
        return $permalink;
    }

    protected static function PrintStackTrace(): void
    {
        $e = new \Exception;
        var_dump($e->getTraceAsString());
    }

    /**
     * @param WP_Post $post
     * @return array|mixed|object|WP_Error|null
     */
    private static function GetLastCategory($post)
    {
        // See the problems when a post is under more categories
        // https://stackoverflow.com/questions/45693271/getting-wordpress-single-post-primary-category

        // Find the canonical category for the post (assigned category with
        // lowest id)
        $cats = get_the_category($post->ID);
        if ($cats)
        {
            $cats = wp_list_sort($cats, 'term_id', 'asc');
            $category = $cats[0];
        }
        else
        {
            $category = get_category(absint(get_option('default_category')));
        }
        return $category;
    }

    /**
     * There is not primary category in wordpress if you have installed Yoast SEO plugin then a new feature will be appear on Single Posts category selection in admin area in order to choose primary category .
     * @param WP_Post $post
     * @param string $term
     * @return array|mixed|object|WP_Error|null
     */
    private static function get_post_primary_category($post, $term = 'category')
    {
        if (class_exists('WPSEO_Primary_Term'))
        {
            // Show Primary category by Yoast if it is enabled & set
            $wpseo_primary_term = new WPSEO_Primary_Term($term, $post->ID);
            $primary_term = get_term($wpseo_primary_term->get_primary_term());

            if (!is_wp_error($primary_term))
            {
                //PC::debug($primary_term);
                return $primary_term;
            }
        }

        if (empty($return['primary_category']))
        {
            //PC::debug("primary_category è vuota");
            return self::GetLastCategory($post);
        }
    }

    /**
     * Redirects fully hierarchical category links to the single category link.
     *
     * @since 2.0
     */
    public static function template_redirect()
    {
        global $wp_query, $post;

        $redirect = null;
        $category_name = isset($wp_query->query['category_name']) ? $wp_query->query['category_name'] : '';

        if (is_category())
        {
            if ($category_name && $category_name != $wp_query->query_vars['category_name'])
            {
                $redirect = self::category_link('', $wp_query->query_vars['cat']);
            }
        }
        elseif (is_single())
        {
            //PC::debug($category_name);

            if ($category_name && substr_count($category_name, '/') >= 1)
            {
                $redirect = get_permalink($post);
            }
        }

        if ($redirect)
        {
            wp_redirect($redirect, self::get_http_redirect_status());
        }
    }

    /**
     * Returns category URI for a given category.
     *
     * If the given category is hierarchical, then this function kicks into gear to
     * reduce a hierarchical category structure to its lowest category in the link.
     *
     * @param string $catlink The default URI for the category
     * @param int $category_id The category ID
     * @param string $taxonomy Taxonomy slug.
     * @return string The category URI
     */
    public static function category_link($catlink, $category_id, $taxonomy)
    {
        global $wp_rewrite;

        // Bail early if taxonomy is not 'category'.
        if ('category' !== $taxonomy)
        {
            return $catlink;
        }

        $catlink = $wp_rewrite->get_category_permastruct();

        if (!$catlink)
        {
            $file = trailingslashit(get_option('siteurl'));
            $catlink = $file . '?cat=' . $category_id;
        }
        else
        {
            $category = get_category($category_id);
            if (is_wp_error($category))
            {
                return $category;
            }
            $category_nicename = $category->slug;

            //$catlink = str_replace('/category/', '/', $catlink);
            $catlink = str_replace('%category%', $category_nicename, $catlink);
            $catlink = home_url(user_trailingslashit($catlink, 'category'));
        }

        return $catlink;
    }

    /**
     * Returns the HTTP status to use for redirects.
     *
     * @return string
     * @uses filter c2c_single_category_redirect_code
     *
     * @since 2.2
     */
    public static function get_http_redirect_status()
    {
        /**
         * Filters the HTTP status code used for redirects.
         *
         * @param int The HTTP status code to be used for redirects. Default 301.
         * @since 2.0
         *
         */
        return (int)apply_filters('c2c_single_category_redirect_status', 301);
    }

} // end c2c_SingleCategoryPermalink

add_action('plugins_loaded', array('c2c_SingleCategoryPermalink', 'init'), 1);